File: backend/app/schemas/recommendation.py
import uuid
from datetime import datetime

from pydantic import BaseModel, Field

from app.calculations.models import CalculationResult


class RecommendationRequest(BaseModel):
    """Formulaire 'Nouveau calcul' (section 11). Les 5 premiers champs sont obligatoires."""

    project_id: int | None = None

    required_flux_lm: float = Field(..., gt=0, description="Flux lumineux demande (lm)")
    max_power_w: float = Field(..., gt=0, description="Puissance totale maximale autorisee (W)")
    required_cct_k: int = Field(..., gt=0, description="Temperature de couleur demandee (K)")
    voltage_nominal_v: float = Field(..., gt=0, description="Tension nominale du module (V)")
    current_nominal_ma: float = Field(..., gt=0, description="Courant nominal (mA)")

    protocol: str | None = Field(None, description="DALI, DALI-2, D4i, 0-10V ou 1-10V")
    led_package: str | None = None
    road_type: str | None = None
    pole_height_m: float | None = Field(None, gt=0)
    pole_spacing_m: float | None = Field(None, gt=0)
    ambient_temperature_c: float | None = None

    # --- Calculateur technique (geometrie routiere / implantation / energie) ---
    road_width_m: float | None = Field(None, gt=0)
    road_length_m: float | None = Field(None, gt=0)
    layout_type: str | None = Field(None, description="unilateral, opposite, staggered ou central")
    operating_hours_per_year: float | None = Field(None, gt=0)
    energy_price_per_kwh: float | None = Field(None, gt=0)


class ComponentRef(BaseModel):
    id: int
    manufacturer: str
    reference: str


class ScoresOut(BaseModel):
    electrical: float
    photometric: float
    mechanical: float
    thermal: float
    data_quality: float


class EvidenceOut(BaseModel):
    """Une preuve documentaire retrouvee pour une configuration retenue (V2 - RAG)."""

    category: str
    document: str
    section: str | None
    page: int | None
    relevance_score: float
    summary: str


class DocumentaryAnalysisOut(BaseModel):
    """Synthese documentaire d'une configuration (V2). Ne modifie jamais `overall_score`.

    `confidence` reste distincte de la compatibilite technique : une confiance
    documentaire basse ne signifie jamais qu'une configuration est incompatible.
    """

    confidence: str  # high / medium / low / insufficient_evidence
    evidence_count: int
    evidence: list[EvidenceOut]
    missing_evidence: list[str]


class RecommendationItem(BaseModel):
    id: int
    rank: int
    overall_score: float
    driver: ComponentRef
    module: ComponentRef
    lens: ComponentRef | None
    scores: ScoresOut
    validated_rules: list[str]
    warnings: list[str]
    blocking_reasons: list[str]
    explanation: str
    validation_status: str | None = None
    documentary_analysis: DocumentaryAnalysisOut | None = None
    technical_calculations: CalculationResult | None = None


class RejectedSummary(BaseModel):
    drivers_rejected: int
    modules_rejected: int
    lenses_rejected: int


class RecommendationResponse(BaseModel):
    status: str
    request_id: uuid.UUID
    run_id: int
    message: str
    recommendations: list[RecommendationItem]
    rejected_summary: RejectedSummary
    blocking_reasons: list[str]
    suggestions: list[str]
    created_at: datetime


class ValidateResultRequest(BaseModel):
    """Body de POST /api/recommendation-results/{result_id}/validate|reject.

    `validator_name` est obligatoire : une configuration validee sans nom de
    consultant ne doit jamais pouvoir produire un rapport PDF final.
    """

    validator_name: str = Field(..., min_length=1, description="Nom du consultant qui valide/rejette")
    comment: str | None = None

File: frontend/src/schemas/recommendation.ts
import { z } from "zod";

export const recommendationFormSchema = z.object({
  required_flux_lm: z
    .number({ error: "Le flux lumineux est obligatoire." })
    .positive("Le flux lumineux doit etre superieur a 0."),
  max_power_w: z
    .number({ error: "La puissance maximale est obligatoire." })
    .positive("La puissance maximale doit etre superieure a 0."),
  required_cct_k: z
    .number({ error: "La temperature de couleur (CCT) est obligatoire." })
    .positive("La CCT doit etre superieure a 0."),
  voltage_nominal_v: z
    .number({ error: "La tension nominale est obligatoire." })
    .positive("La tension doit etre superieure a 0."),
  current_nominal_ma: z
    .number({ error: "Le courant nominal est obligatoire." })
    .positive("Le courant doit etre superieur a 0."),

  protocol: z.string().optional(),
  led_package: z.string().optional(),
  road_type: z.string().optional(),
  pole_height_m: z.number().positive("La hauteur doit etre positive.").optional(),
  pole_spacing_m: z.number().positive("L'espacement doit etre positif.").optional(),
  ambient_temperature_c: z.number().optional(),

  // Champs additionnels du calculateur technique (jamais envoyes a createRecommendation()) :
  // alimentent uniquement l'apercu POST /api/calculations/preview.
  road_width_m: z.number().positive("La largeur doit etre positive.").optional(),
  road_length_m: z.number().positive("La longueur doit etre positive.").optional(),
  // z.string() (et non z.enum(...)) : le <select> renvoie "" par defaut, que z.enum().optional()
  // rejette (seuls undefined ou une valeur de l'enum sont acceptes) — meme choix que `protocol` ci-dessus.
  layout_type: z.string().optional(),
  operating_hours_per_year: z.number().positive("Les heures doivent etre positives.").optional(),
  energy_price_per_kwh: z.number().positive("Le tarif doit etre positif.").optional(),
});

export type RecommendationFormValues = z.infer<typeof recommendationFormSchema>;

/** Convertit la valeur brute d'un <input type="number"> : chaine vide -> undefined (jamais NaN). */
export function setValueAsNumber(value: string) {
  return value === "" ? undefined : Number(value);
}
